v2/internal/frontend/assetserver/assetserver_common_dev.go

```diff
@@ -0,0 +1,57 @@
+//go:build dev
+// +build dev
+
+package assetserver
+
+import (
+	"context"
+	"embed"
+	"io/fs"
+	"os"
+	"path/filepath"
+
+	"github.com/wailsapp/wails/v2/internal/logger"
+)
+
+func isAssetDirOverride(ctx context.Context, serverType string, assets fs.FS) (fs.FS, error) {
+	_logger, _ := ctx.Value("logger").(*logger.Logger)
+	logWarning := func(msg string) {
+		if _logger == nil {
+			return
+		}
+
+		_logger.Warning("[%s] Reloading changed asset files won't work: %s", serverType, msg)
+	}
+
+	_, isEmbedFs := assets.(embed.FS)
+	if assetdir := ctx.Value("assetdir"); assetdir == nil {
+		// We are in dev mode, but have no assetdir, let's check the type of the assets FS.
+		if isEmbedFs {
+			logWarning("A 'embed.FS' has been provided, but no assetdir has been defined")
+		} else {
+			// That's fine, the user is using another fs.FS and also has not set an assetdir
+		}
+	} else {
+		// We are in dev mode and an assetdir has been defined
+		if isEmbedFs {
+			// If the fs.FS is an embed.FS we assume it's serving the files from this directory, therefore replace the assets
+			// fs.FS with this directory.
+			// Are there any better ways to detect this? If we could somehow determine the compile time source path for the
+			// embed.FS this would be great. Currently there seems no way to get to this information.
```

Comment:
I also thought about removing the assetdir flag during the implementation of the PR, and I would love it if we could remove it somehow.

It should work as long as the `embed.FS` variable has been defined on the main package, like it's done during a new project initialisation. The content to be included into the `embed.FS` is always relative to the package where the variable has been defined. If someone puts the variable into a sub package, let's say into the frontend package, we would only know the relative path `src` and we wouldn't know the relative path into the frontend package. In this case we would construct an absolute path of `project root` joined with `src`.

Reply:
Considering the frontend is *always* in projectroot/frontend, I think it would be easy to walk the embed.FS and see if it lines up, maybe check if the files exist in both places?

Reply:
Yes, that's a good idea. We could determine the path to index.html in the embed.FS and see if `project root` joined with that relative path exists on the filesystem and also contains an index.html. And if it not lines up, we could log a warning. Should we leave the assetdir as a fallback or would you like to completely remove it?

Reply:
Initially fallback otherwise, if it doesn't line up you'll have problems

Reply:
I'm going to update the PR tomorrow with these changes.

Reply:
@leaanthony I've updated the PR, please let me know what you think.